### PR TITLE
[CHORE] considering global limits over enforced

### PR DESCRIPTION
**Pull request by nicolastakashi**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Considering global limits over enforced limits when no user limits are set.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [X] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
considering global limits over enforced limits when no user limits are set.
```

### CODE REVIEW

**Comment by nicolastakashi** [comment]
Leveraging global limits instead of enforced limits if no user limit is set as we discussed at #6589

**Comment by simonpasquier** [comment]
Another option would be 
1. Always set `.spec.sampleLimit` to `.spec.enforcedSampleLimit` when it the former is nil/zero and the latter > 0.
2. Don't set per-scrape sample limit in the generated config if the value is nil/zero (and rely on the global limit being always set).

**Comment by nicolastakashi** [comment]
> Another option would be
> 
> 1. Always set `.spec.sampleLimit` to `.spec.enforcedSampleLimit` when it the former is nil/zero and the latter > 0.
> 2. Don't set per-scrape sample limit in the generated config if the value is nil/zero (and rely on the global limit being always set).

I'm Ok with that!
Is it ok to have the same behaviour for all limits?

**Comment by simonpasquier** [comment]
>  Is it ok to have the same behaviour for all limits?

Yes it definitely makes sense to be consistent here. Though I assume that users are mostly concerned about sample limits.

**Comment by nicolastakashi** [comment]
@simonpasquier PTAL 🙏🏽

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -2257,29 +2256,39 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
...
-	if cpf.KeepDroppedTargets != nil {
-		slice = cg.WithMinimumVersion("2.47.0").AppendMapItem(slice, "keep_dropped_targets", *cpf.KeepDroppedTargets)
-	}
+
+	slice = cg.appendGlobalLimits(slice, "2.45.0", "sample_limit", cpf.SampleLimit, cpf.EnforcedSampleLimit)
+	slice = cg.appendGlobalLimits(slice, "2.45.0", "target_limit", cpf.TargetLimit, cpf.EnforcedTargetLimit)
+	slice = cg.appendGlobalLimits(slice, "2.45.0", "label_limit", cpf.LabelLimit, cpf.EnforcedLabelLimit)
+	slice = cg.appendGlobalLimits(slice, "2.45.0", "label_name_length_limit", cpf.LabelNameLengthLimit, cpf.EnforcedLabelNameLengthLimit)
+	slice = cg.appendGlobalLimits(slice, "2.45.0", "label_value_length_limit", cpf.LabelValueLengthLimit, cpf.EnforcedLabelValueLengthLimit)
+	slice = cg.appendGlobalLimits(slice, "2.47.0", "keep_dropped_targets", cpf.KeepDroppedTargets, cpf.EnforcedKeepDroppedTargets)
```
Comment (ArthurSens) [inline pkg/prometheus/promcfg.go:2291]: Nice! Looks a lot cleaner :)

File: pkg/prometheus/testdata/KeepDroppedTargets.golden [hunk 2]
```diff
@@ -4,6 +4,7 @@ global:
   external_labels:
     prometheus: default/test
     prometheus_replica: $(POD_NAME)
+  keep_dropped_targets: 1000
```
Comment (ArthurSens) [inline pkg/prometheus/testdata/KeepDroppedTargets.golden:7]: Was this change intentional? Do we want to make the same changes to `keep_dropped_targets`?

**Review by ArthurSens — COMMENTED** [review]
It generally looks good to me, but almost all golden files might need to be renamed.

I think it's also worth updating API docs for all enforced fields, clarifying the expected behavior.

**Comment by ArthurSens** [comment]
This is definitely a breaking change though. 

You can see the changes being made to golden files where `enforced.*limits` were declared but limits were not declared in the scrape objects. All of a sudden a new limit is added to the configuration.

If we proceed with this, updating prometheus-operator to v0.75 can become quite troublesome to some users 😕

File: pkg/prometheus/testdata/KeepDroppedTargets.golden (hunk 2 above)
Comment (nicolastakashi) [reply pkg/prometheus/testdata/KeepDroppedTargets.golden:7]: From what me and @simonpasquier discussed, this should be a consistent behaviour through all the limits.

**Comment by nicolastakashi** [comment]
> This is definitely a breaking change though.
> 
> You can see the changes being made to golden files where `enforced.*limits` were declared but limits were not declared in the scrape objects. All of a sudden a new limit is added to the configuration.
> 
> If we proceed with this, updating prometheus-operator to v0.75 can become quite troublesome to some users 😕

I'm not sure if this is a breaking change, because the behavior is still the same in the end when you're using enforcedSampleLimits the only difference is, instead of setting the sample limit on the scrape config it will be placed on the global config.

wdyt @simonpasquier ?

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -1617,7 +1616,7 @@ func getLimit(user *uint64, enforced *uint64) *uint64 {
 	}
 
 	if user == nil {
-		return enforced
+		return user
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1619]: I'm confused by this change.

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2257,29 +2256,39 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
 	return append(slice, yaml.MapItem{Key: "evaluation_interval", Value: evaluationInterval})
 }
 
+func (cg *ConfigGenerator) appendGlobalLimits(slice yaml.MapSlice, version string, limitKey string, limit *uint64, enforcedLimit *uint64) yaml.MapSlice {
+	if limit == nil && enforcedLimit != nil && *enforcedLimit > 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2260]: (suggestion) and you can get rid of the next block IIUC
```suggestion
	if ptr.Deref(limit, 0) == 0 && ptr.Deref(enforcedLimit, 0) > 0 {
```

File: pkg/prometheus/promcfg.go [hunk 5]
```diff
@@ -2257,29 +2256,39 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
...
+		slice = cg.WithMinimumVersion(version).AppendMapItem(slice, limitKey, *enforcedLimit)
+		return slice
+	}
+
+	if enforcedLimit != nil && *enforcedLimit > 0 && limit != nil && *limit < 1 {
+		slice = cg.WithMinimumVersion(version).AppendMapItem(slice, limitKey, *enforcedLimit)
+	}
+
+	if limit != nil && *limit > 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2269]:
```suggestion
	if ptr.Deref(limit, 0) > 0 {
```

**Comment by simonpasquier** [comment]
Technically it's a change so it should be highlighted as such in the changelog. Today if you configure both `.spec.sampleLimit` and `.spec.enforcedSampleLimit` in the Prometheus object, only the latter is taken into account so I wouldn't expect users to define both fields in practice. If you've configure only one of them, there should be no visible change.
So my position is that the new behavior makes more sense and it should impact a small fraction of users (if any).

> I think it's also worth updating API docs for all enforced fields, clarifying the expected behavior.

Agreed.

File: pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden [hunk 6]
```diff
@@ -66,4 +66,3 @@ scrape_configs:
     - __tmp_hash
     regex: $(SHARD)
     action: keep
-  target_limit: 1000
```
Comment (simonpasquier) [inline pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden:69]: hmm we had a bug?

File: pkg/prometheus/testdata/LabelLimits_NoLimit_v2.27.0_enforceLimit1000.golden [hunk 7]
```diff
@@ -66,4 +66,3 @@ scrape_configs:
     - __tmp_hash
     regex: $(SHARD)
     action: keep
-  label_limit: 1000
```
Comment (simonpasquier) [inline pkg/prometheus/testdata/LabelLimits_NoLimit_v2.27.0_enforceLimit1000.golden:69]: is this correct?

File: pkg/prometheus/promcfg.go (hunk 3 above)
Comment (nicolastakashi) [reply pkg/prometheus/promcfg.go]: I just pushed another version, can you take a look?

File: pkg/prometheus/promcfg.go [hunk 8]
```diff
@@ -1050,7 +1050,6 @@ func (cg *ConfigGenerator) generatePodMonitorConfig(
 
 	relabelings = generateAddressShardingRelabelingRules(relabelings, shards)
 	cfg = append(cfg, yaml.MapItem{Key: "relabel_configs", Value: relabelings})
-
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1053]: (nit) can we keep the empty line?

File: pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden (hunk 6 above)
Comment (simonpasquier) [reply pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden:69]: @nicolastakashi thoughts?

**Review by simonpasquier — COMMENTED** [review]
I believe that we have an issue to fix depending on Prometheus versions... IIUC the global limits require 2.45.0 while scrape limits where are available since 2.21.0 & 2.27.0.

I'm afraid that we'll have to set the limits on the individual scrape configs rather than relying on global settings to accommodate older versions.

File: pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden (hunk 6 above)
Comment (nicolastakashi) [reply pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden:69]: @simonpasquier checking yet, I'll update soon.

**Comment by simonpasquier** [comment]
Hmm enforcedSampleLimits worked previously with versions <  2.45 so it's a bit extreme to say that it doesn't anymore.

File: pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden (hunk 6 above)
Comment (simonpasquier) [reply pkg/prometheus/testdata/TargetLimits-1_Versionv2.21.0_Enforce1000.golden:69]: I suppose that it's the isssue with global sample_limit not being supported before 2.45.

**Comment by nicolastakashi** [comment]
> Hmm enforcedSampleLimits worked previously with versions < 2.45 so it's a bit extreme to say that it doesn't anymore.
@simonpasquier I'm keeping the current behavior is version is < 2.45, so we don't introduce any breaking change on this case, wdyt?

File: pkg/prometheus/promcfg.go [hunk 9]
```diff
@@ -1611,16 +1615,16 @@ func generateRunningFilter() yaml.MapSlice {
...
 		return user
 	}
 
-	if user == nil {
+	if isGlobalLimitNotSupported && user == nil {
 		return enforced
 	}
 
-	if *enforced > *user {
+	if ptr.Deref(user, 0) <= ptr.Deref(enforced, 0) {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1627]: (nit)
```suggestion
	if ptr.Deref(enforced, 0) > ptr.Deref(user, 0) {
```

File: pkg/prometheus/promcfg.go [hunk 10]
```diff
@@ -2257,29 +2261,35 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
 	return append(slice, yaml.MapItem{Key: "evaluation_interval", Value: evaluationInterval})
 }
 
+func (cg *ConfigGenerator) appendGlobalLimits(slice yaml.MapSlice, limitKey string, limit *uint64, enforcedLimit *uint64) yaml.MapSlice {
+	if ptr.Deref(limit, 0) == 0 && ptr.Deref(enforcedLimit, 0) > 0 {
+		slice = cg.AppendMapItem(slice, limitKey, *enforcedLimit)
+		return slice
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2267]: (nit)
```suggestion
		return cg.AppendMapItem(slice, limitKey, *enforcedLimit)
```

File: pkg/prometheus/promcfg.go [hunk 11]
```diff
@@ -2257,29 +2261,35 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
 	return append(slice, yaml.MapItem{Key: "evaluation_interval", Value: evaluationInterval})
 }
 
+func (cg *ConfigGenerator) appendGlobalLimits(slice yaml.MapSlice, limitKey string, limit *uint64, enforcedLimit *uint64) yaml.MapSlice {
+	if ptr.Deref(limit, 0) == 0 && ptr.Deref(enforcedLimit, 0) > 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2265]:
```suggestion
	// Use the enforced limit if no global limit is defined to ensure that scrape jobs without an explicit limit inherit the enforced limit value.
	if ptr.Deref(limit, 0) == 0 && ptr.Deref(enforcedLimit, 0) > 0 {
```

File: pkg/prometheus/promcfg.go [hunk 12]
```diff
@@ -2257,29 +2261,35 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
...
+		slice = cg.AppendMapItem(slice, limitKey, *enforcedLimit)
+		return slice
+	}
+
+	if ptr.Deref(limit, 0) > 0 {
+		slice = cg.AppendMapItem(slice, limitKey, *limit)
+		return slice
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2272]:
```suggestion
		return cg.AppendMapItem(slice, limitKey, *limit)
```

File: pkg/prometheus/promcfg.go [hunk 13]
```diff
@@ -1611,16 +1615,16 @@ func generateRunningFilter() yaml.MapSlice {
...
 }
 
-func getLimit(user *uint64, enforced *uint64) *uint64 {
+func getLimit(user *uint64, enforced *uint64, isGlobalLimitNotSupported bool) *uint64 {
 	if enforced == nil {
 		return user
 	}
 
-	if user == nil {
+	if isGlobalLimitNotSupported && user == nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1623]: I'd find it easier to have the version inline here
```suggestion
	// If the Prometheus version doesn't support global limit 
	if cg.version.LT(semver.MustParse("2.45.0")) && user == nil {
```

File: pkg/prometheus/promcfg.go [hunk 14]
```diff
@@ -1611,16 +1615,16 @@ func generateRunningFilter() yaml.MapSlice {
...
 		return enforced
 	}
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1625]: My understanding is that when Prometheus supports global limits then we don't want to set scrape limit == enforced.
I find it also easier to have the version inline here.

```suggestion
	// Prometheus >= 2.45.0 supports global limits in which case the global limit gets applied for all targets that don't specify an explicit limit.
	if cg.version.GTE(semver.MustParse("2.45.0")) && ptr.Deref(user, 0) == 0 {
		return nil
	}
```

File: pkg/prometheus/promcfg_test.go [hunk 15]
```diff
@@ -3321,33 +3321,49 @@ func TestTrackTimestampsStaleness(t *testing.T) {
...
 			enforcedLimit: 1000,
 			limit:         500,
 			golden:        "SampleLimits_Limit500.golden",
 		},
+		{
+			globalLimit:   1000,
+			enforcedLimit: 2000,
+			limit:         -1,
+			golden:        "SampleLimits_GlobalLimit1000_Enforce2000.golden",
+		},
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:3358]: I wonder if we don't need more test cases? In particular with version < 2.4.50?

**Review by simonpasquier — COMMENTED** [review]
Can you also update the API documentation to clarify the behavior change?

File: pkg/prometheus/promcfg.go [hunk 16]
```diff
@@ -2257,29 +2258,34 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
...
 
+func (cg *ConfigGenerator) appendGlobalLimits(slice yaml.MapSlice, limitKey string, limit *uint64, enforcedLimit *uint64) yaml.MapSlice {
+	// Use the enforced limit if no global limit is defined to ensure that scrape jobs without an explicit limit inherit the enforced limit value.
+	if ptr.Deref(limit, 0) == 0 && ptr.Deref(enforcedLimit, 0) > 0 {
+		return cg.AppendMapItem(slice, limitKey, *enforcedLimit)
+	}
+
+	if ptr.Deref(limit, 0) > 0 {
+		return cg.AppendMapItem(slice, limitKey, *limit)
+	}
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2269]: It reads more easily for me. Though it brings the question of what to do when global limit > enforced limit? I suggest that we clamp down the global limit to the enforced value in this case (CEL could help later on to validate this at the API level).
```suggestion
	if ptr.Deref(limit, 0) > 0 {
		return cg.AppendMapItem(slice, limitKey, *limit)
	}

	// <explain why enforcedlimit is used>
	if ptr.Deref(enforcedLimit, 0) > 0 {
		return cg.AppendMapItem(slice, limitKey, *enforcedLimit)
	}
```

File: pkg/prometheus/promcfg.go [hunk 17]
```diff
@@ -1611,16 +1611,17 @@ func generateRunningFilter() yaml.MapSlice {
 	}
 }
 
-func getLimit(user *uint64, enforced *uint64) *uint64 {
+func (cg *ConfigGenerator) getLimit(user *uint64, enforced *uint64) *uint64 {
 	if enforced == nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1615]: (suggestion) easier to use ptr.Deref() from the start?
```suggestion
	if ptr.Deref(enforced, 0) == 0 {
```

File: pkg/prometheus/promcfg.go [hunk 18]
```diff
@@ -1611,16 +1611,17 @@ func generateRunningFilter() yaml.MapSlice {
...
 		return user
 	}
 
-	if user == nil {
+	// If the Prometheus version doesn't support global limit
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1619]: Would it be easier to follow with this logic?
* if *user == 0
  * if version >= 2.45.0, return nil (global limit applies)
  * otherwise return enforced

**Review by simonpasquier — COMMENTED** [review]
I don't remember if I said it already but I think that the new behavior needs to be documented in the API.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 19]
```diff
@@ -514,6 +514,9 @@ type CommonPrometheusFields struct {
 	// It is meant to be used by admins to keep the overall number of
 	// samples/series under a desired limit.
 	//
+	// Starting from Prometheus v2.45.0, this field is used to define global limits if `spec.sampleLimit` is not set.
+	// Admins can combine this field with `spec.sampleLimit` to provide a global upper limit and a lower limit for individual objects.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:518]: It's still a bit unclear IMHO. What if we explain in details the decision tree?

```
When both `enforcedSampleLimit` and `sampleLimit` are defined and greater than zero, the following rules apply:
* Scrape objets without a defined sampleLimit value will inherit the global sampleLimit value (Prometheus <= 2.45.0) or the enforcedSampleLimit value (Prometheus > v2.45.0).
* Scrape objets with a sampleLimit value less than or equal to enforcedSampleLimit keep their specific value.
* Scrape objets with a sampleLimit value greater than enforcedSampleLimit are set to enforcedSampleLimit.
```

From my POV, it's ok to have the detailed documentation for 1 field (e.g. enforced sample limit) and have a reference (e.g. "See `enforcedSampleLimit` to understand the interactions with `<field>`."). It might also good to have a reference to the enforced fields in each global field.

File: pkg/prometheus/promcfg.go [hunk 20]
```diff
@@ -1611,16 +1611,20 @@ func generateRunningFilter() yaml.MapSlice {
...
-func getLimit(user *uint64, enforced *uint64) *uint64 {
-	if enforced == nil {
+func (cg *ConfigGenerator) getLimit(user *uint64, enforced *uint64) *uint64 {
+	if ptr.Deref(enforced, 0) == 0 {
 		return user
 	}
 
-	if user == nil {
+	if ptr.Deref(user, 0) == 0 {
+		// Prometheus 2.45.0 introduced the ability to use global sample_limit
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1620]:
```suggestion
		// With Prometheus >= 2.45.0, the limit value in the global section will always apply, hence there's no need to set the value explicitly.
```

File: pkg/prometheus/promcfg.go [hunk 21]
```diff
@@ -2257,29 +2261,34 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
 	return append(slice, yaml.MapItem{Key: "evaluation_interval", Value: evaluationInterval})
 }
 
+func (cg *ConfigGenerator) appendGlobalLimits(slice yaml.MapSlice, limitKey string, limit *uint64, enforcedLimit *uint64) yaml.MapSlice {
+	if ptr.Deref(limit, 0) > 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2265]: should we check that limit <= enforced? Otheriwse is there a risk of the "soft" limit being higher than the enforced limit?

Comment (nicolastakashi) [reply pkg/prometheus/promcfg.go:2265]: @simonpasquier wdyt?

File: pkg/prometheus/testdata/SampleLimits_GlobalLimit1000_Enforce2000.golden [hunk 22]
```diff
@@ -0,0 +1,69 @@
+global:
+  evaluation_interval: 30s
+  scrape_interval: 30s
+  external_labels:
+    prometheus: default/test
+    prometheus_replica: $(POD_NAME)
+  sample_limit: 2000
```
Comment (simonpasquier) [inline pkg/prometheus/testdata/SampleLimits_GlobalLimit1000_Enforce2000.golden:7]: shouldn't it be 1,000?

File: pkg/prometheus/promcfg.go [hunk 23]
```diff
@@ -2289,29 +2293,38 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
...
+		if ptr.Deref(enforcedLimit, 0) > 0 && *limit < *enforcedLimit {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2298]: shouldn't it be `*limit > *enforcedLimit`?

Comment (nicolastakashi) [reply pkg/prometheus/promcfg.go:2298]: Don't think so!
Sample Limit == Soft Limit

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:2298]: I'd rather expect the opposite: enforced (hard) limits should always be greater than global (soft) limits?
In any case it should be documented in the `sampleLimits` field.

**Comment by nicolastakashi** [comment]
@ArthurSens @simonpasquier PTAL

File: pkg/prometheus/promcfg.go [hunk 24]
```diff
@@ -2289,29 +2293,38 @@ func (cg *ConfigGenerator) appendEvaluationInterval(slice yaml.MapSlice, evaluat
 	return append(slice, yaml.MapItem{Key: "evaluation_interval", Value: evaluationInterval})
 }
 
+func (cg *ConfigGenerator) appendGlobalLimits(slice yaml.MapSlice, limitKey string, limit *uint64, enforcedLimit *uint64) yaml.MapSlice {
+	if ptr.Deref(limit, 0) > 0 {
+		if ptr.Deref(enforcedLimit, 0) > 0 && *limit > *enforcedLimit {
+			level.Warn(cg.logger).Log("msg", limitKey, "limit is greather than enforced limit, using enforced limit")
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2299]:
```suggestion
			level.Warn(cg.logger).Log("msg", fmt.Sprintf("%q is greater than the enforced limit, using enforced limit", limitKey), "limit", *limit, "enforced_limit", *enforcedLimit)
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 25]
```diff
@@ -514,6 +514,12 @@ type CommonPrometheusFields struct {
 	// It is meant to be used by admins to keep the overall number of
 	// samples/series under a desired limit.
 	//
+	// When both `enforcedSampleLimit` and `sampleLimit` are defined and greater than zero, the following rules apply:
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:517]: shall we say that when sampleLimit > enforcedSampleLimit, we consider only enforcedSampleLimit?

Comment (nicolastakashi) [reply pkg/apis/monitoring/v1/prometheus_types.go:517]: its already there I think.

```If Prometheus version is >= 2.45.0 and the `enforcedTargetLimit` is greater than the `targetLimit`, the `targetLimit` will be set to `enforcedTargetLimit`.```

Comment (simonpasquier) [reply pkg/apis/monitoring/v1/prometheus_types.go:517]: :+1:

**Review by simonpasquier — APPROVED** [review]
Thanks! :tada:
